test/e2e/network/endpointslice.go

```diff
@@ -0,0 +1,428 @@
+		framework.ExpectNoError(err, "Error fetching EndpointSlice for %s/%s Service", ns, svcName)
+		if len(esList.Items) == 0 {
+			framework.Logf("EndpointSlice for %s/%s Service not found", ns, svcName)
+			continue
+		}
+		if len(esList.Items) != numSlices {
+			framework.Logf("Expected %d EndpointSlices for %s/%s Service, got %d", numSlices, ns, svcName, len(esList.Items))
+			continue
+		}
+
+		actualNumEndpoints := 0
+		for _, endpointSlice := range esList.Items {
+			actualNumEndpoints += len(endpointSlice.Endpoints)
+		}
+		if actualNumEndpoints != numEndpoints {
+			framework.Logf("EndpointSlices for %s/%s Service have %d/%d endpoints", ns, svcName, actualNumEndpoints, numEndpoints)
+			continue
+		}
+		return esList.Items
+	}
+
+	framework.Failf("failed to get EndpointSlices for %s/%s Service", ns, svcName)
+	return []discovery.EndpointSlice{}
+}
+
+// waitForEndpointWithNumEndpoints waits for Endpoints for the specified Service
+// name to be ready with the expected number of IPs.
+func waitForMatchingEndpoints(cs clientset.Interface, ns, svcName string, numIPs, numSubsets int) *v1.Endpoints {
+	// registerTimeout is how long to wait for an Endpoint to be registered.
+	registerTimeout := time.Minute
+
+	for t := time.Now(); time.Since(t) < registerTimeout; time.Sleep(framework.Poll) {
+		endpoints, err := cs.CoreV1().Endpoints(ns).Get(svcName, metav1.GetOptions{})
+		if err != nil {
+			if errors.IsNotFound(err) {
+				framework.Logf("Endpoints for %s/%s Service not found", ns, svcName)
+				continue
+			} else {
+				framework.ExpectNoError(err, "Error fetching Endpoints for %s/%s Service", ns, svcName)
+			}
+		}
+		if len(endpoints.Subsets) != numSubsets {
+			framework.Logf("Endpoints for %s/%s Service with %d/%d Subsets", ns, svcName, len(endpoints.Subsets), numSubsets)
+			continue
+		}
+
+		actualNumIPs := 0
+		for _, endpointSubset := range endpoints.Subsets {
+			actualNumIPs += len(endpointSubset.Addresses) + len(endpointSubset.NotReadyAddresses)
+		}
+		if actualNumIPs != numIPs {
+			framework.Logf("Endpoints for %s/%s Service with %d/%d IPs", ns, svcName, actualNumIPs, numIPs)
+			continue
+		}
+		return endpoints
+	}
+
+	framework.Failf("failed to get Endpoints for %s/%s Service", ns, svcName)
+	return &v1.Endpoints{}
+}
+
+// waitForPodWithIP waits for a Pod matching a given namespace and name to be
+// assigned an IP and returns that Pod.
+func waitForPodWithIP(cs clientset.Interface, ns, podName string) *v1.Pod {
+	// registerTimeout is how long to wait for Pod to have IP.
+	registerTimeout := 3 * time.Minute
+
+	for t := time.Now(); time.Since(t) < registerTimeout; time.Sleep(framework.Poll) {
+		pod, err := cs.CoreV1().Pods(ns).Get(podName, metav1.GetOptions{})
+		if err != nil {
+			if errors.IsNotFound(err) {
+				framework.Logf("Pod %s/%s not found", ns, podName)
+				continue
+			} else {
+				framework.ExpectNoError(err, "Error fetching Pod %s/%s: %v", ns, podName, err)
+			}
+		}
+
+		if pod.Status.PodIP == "" {
```

Comment:
You can check pod readiness instead. Probably more accurate: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/common/pods.go#L803

Reply:
Good call, thanks!